File: bundles/org.eclipse.rap.rwt/js/org/eclipse/rap/rwt/widgets/Spinner.js
/*******************************************************************************
 * Copyright (c) 2007 Innoopract Informationssysteme GmbH.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 * 
 * Contributors:
 *     Innoopract Informationssysteme GmbH - initial API and implementation
 ******************************************************************************/

qx.OO.defineClass(
  "org.eclipse.rap.rwt.widgets.Spinner", 
  qx.ui.form.Spinner,
  function( readOnly, border ) {
    qx.ui.form.Spinner.call( this );
    this._textfield.setReadOnly( readOnly );
    this._isModified = false;
    this._readOnly = readOnly;
    if( border ) {
      this.addState( "rwt_BORDER" );
    }
    else {
      this._upbutton.addState( "rwt_FLAT" );
      this._downbutton.addState( "rwt_FLAT" );
    }
    if( !readOnly ) {
      this._manager.addEventListener( "changeValue", this._onChangeValue, this );
      this._textfield.addEventListener( "keyinput", this._onChangeValue, this );
      this._textfield.addEventListener( "blur", this._sendModifyText, this );
      this.addEventListener( "changeEnabled", this._onChangeEnabled, this );
    }
    this._updateButtonEnablement();
  }
);

qx.OO.addProperty( { name : "hasModifyListener", type : "boolean" } );

qx.Proto.setFont = function( value ) {
  this._textfield.setFont( value );
}

qx.Proto.dispose = function() {
  if( this.getDisposed() ) {
    return;
  }
  if( !this._readOnly ) {
    this._manager.removeEventListener( "changeValue", this._onChangeValue, this );
    this._textfield.removeEventListener( "keyinput", this._onChangeValue, this );
    this._textfield.removeEventListener( "blur", this._sendModifyText, this );
    this.removeEventListener( "changeEnabled", this._onChangeEnabled, this );
  }
  return qx.ui.form.Spinner.prototype.dispose.call( this );
}

/**
 * HACK: override qx.ui.form.Spinner._increment and _pageIncrement to achieve
 * read-only beavior.
 */
qx.Proto._increment = function() {
  if( !this._readOnly ) {
    qx.ui.form.Spinner.prototype._increment.call( this )
  }
}

qx.Proto._pageIncrement = function() {
  if( !this._readOnly ) {
    qx.ui.form.Spinner.prototype._pageIncrement.call( this )
  }
}

qx.Proto._onChangeValue = function( evt ) {
  if( !org_eclipse_rap_rwt_EventUtil_suspend && !this._isModified ) {
    this._isModified = true;
    var req = org.eclipse.rap.rwt.Request.getInstance();
    req.addEventListener( "send", this._onSend, this );
    if( this.getHasModifyListener() ) {
      qx.client.Timer.once( this._sendModifyText, this, 500 );
    }
  }
}

qx.Proto._onChangeEnabled = function( evt ) {
  this._textfield.setEnabled( this.getEnabled() );
  this._updateButtonEnablement();
}

qx.Proto._updateButtonEnablement = function() {
  if( this.getEnabled() ) {
    this._upbutton.setEnabled( !this._readOnly );
    this._downbutton.setEnabled( !this._readOnly );
  } else {
    this._upbutton.setEnabled( false );
    this._downbutton.setEnabled( false );
  }
}

qx.Proto._onSend = function( evt ) {
  this._isModified = false;
  var widgetManager = org.eclipse.rap.rwt.WidgetManager.getInstance();
  var id = widgetManager.findIdByWidget( this );
  var req = org.eclipse.rap.rwt.Request.getInstance();
  req.addParameter( id + ".selection", this.getValue() );
  req.removeEventListener( "send", this._onSend, this );
}

qx.Proto._sendModifyText = function( evt ) {
  if( this._isModified ) {
    var widgetManager = org.eclipse.rap.rwt.WidgetManager.getInstance();
    var id = widgetManager.findIdByWidget( this );
    var req = org.eclipse.rap.rwt.Request.getInstance();
    req.addEvent( "org.eclipse.rap.rwt.events.modifyText", id );
    req.send();
    this._isModified = false;
  }
}

File: bundles/org.eclipse.rap.rwt/src/org/eclipse/rap/rwt/widgets/Spinner.java
/*******************************************************************************
 * Copyright (c) 2002-2006 Innoopract Informationssysteme GmbH.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 * 
 * Contributors:
 *     Innoopract Informationssysteme GmbH - initial API and implementation
 ******************************************************************************/

package org.eclipse.rap.rwt.widgets;

import java.text.*;
import java.util.Locale;
import org.eclipse.rap.rwt.RWT;
import org.eclipse.rap.rwt.graphics.Point;
import org.eclipse.rap.rwt.graphics.Rectangle;
import org.eclipse.rap.rwt.internal.graphics.FontSizeEstimation;
import com.w4t.util.SessionLocale;


// TODO [rh] style WRAP not yet supported
// TODO cut/copy/past not implemented
// TODO SelectionListener: widgetSelected is fired whenever the value changes
public class Spinner extends Composite {
  
  private static final int UP_DOWN_HEIGHT = 18;
  
  private int digits = 0;
  private int increment = 1;
  private int maximum = 100;
  private int minimum = 0;
  private int pageIncrement = 10;
  private int selection = 0;

  public Spinner( final Composite parent, final int style ) {
    super( parent, checkStyle( style ) );
  }

  // TODO [rh] the qooxdoo Spinner widget does not provide decimal places
  public int getDigits () {
    checkWidget ();
    return digits;
  }

//  public void setDigits( final int value ) {
//    checkWidget();
//  }

  /////////////////////////////////////////
  // Methods to control range and increment 
  
  public int getIncrement () {
    checkWidget ();
    return increment;
  }

  public void setIncrement( final int value ) {
    checkWidget();
    if( value >= 1 ) {
      increment = value;
    }
  }
  
  public int getMinimum () {
    checkWidget ();
    return minimum;
  }
  
  public void setMinimum( final int value ) {
    checkWidget();
    if( value >= 0 && value <= maximum ) {
      minimum = value;
      if( selection < minimum ) {
        selection = minimum;
      }
    }
  }
  
  public int getMaximum () {
    checkWidget ();
    return maximum;
  }

  public void setMaximum( final int value ) {
    checkWidget();
    if( value >= 0 && value >= minimum ) {
      maximum = value;
      if( selection > maximum ) {
        selection = maximum;
      }
    } 
  }
  
  public int getPageIncrement () {
    checkWidget ();
    return pageIncrement;
  }
  
  public void setPageIncrement( final int value ) {
    checkWidget();
    if( value >= 1 ) {
      pageIncrement = value;
    } 
  }

  public int getSelection () {
    checkWidget();
    return selection;
  }
  
  
  public void setSelection( final int value ) {
    checkWidget();
    selection = Math.min( Math.max( minimum, value ), maximum );
  }
  
  public void setValues( final int selection,
                         final int minimum,
                         final int maximum,
                         final int digits,
                         final int increment,
                         final int pageIncrement )
  {
    setMinimum( minimum );
    setMaximum( maximum );
    // setDigits( digits ) - ignore since we cannot (yet) handle digits
    setIncrement( increment );
    setPageIncrement( pageIncrement );
    setSelection( selection );
  }

  ///////////////////
  // Size calculation
  
  public Point computeSize( final int wHint,
                            final int hHint,
                            final boolean changed )
  {
    checkWidget();
    int width = 0;
    int height = 0;
    if( wHint == RWT.DEFAULT || hHint == RWT.DEFAULT ) {
      String string = String.valueOf( maximum );
      if( digits > 0 ) {
        StringBuffer buffer = new StringBuffer();
        buffer.append( string );
        buffer.append( getDecimalSeparator() );
        int count = digits - string.length();
        while( count >= 0 ) {
          buffer.append( "0" );
          count--;
        }
        string = buffer.toString();
      }
      Point textSize = FontSizeEstimation.stringExtent( string, getFont() );
      width = textSize.y;
      height = textSize.x;
    }
    if( width == 0 ) {
      width = DEFAULT_WIDTH;
    }
    if( height == 0 ) {
      height = DEFAULT_HEIGHT;
    }
    if( wHint != RWT.DEFAULT ) {
      width = wHint;
    }
    if( hHint != RWT.DEFAULT ) {
      height = hHint;
    }
    Rectangle trim = computeTrim( 0, 0, width, height );
    if( hHint == RWT.DEFAULT ) {
      int upDownHeight = UP_DOWN_HEIGHT + 2 * getBorderWidth();
      trim.height = Math.max( trim.height, upDownHeight );
    }
    return new Point( trim.width, trim.height );
  }

  public Rectangle computeTrim( final int x,
                                final int y,
                                final int width,
                                final int height )
  {
    checkWidget();
    Rectangle result = new Rectangle( x, y, width, height );
    int margins = 1;
    result.x -= margins;
    result.width += margins;
    if( ( style & RWT.BORDER ) != 0 ) {
      result.x -= 1;
      result.y -= 1;
      result.width += 2;
      result.height += 2;
    }
    result.width += ScrollBar.SCROLL_BAR_WIDTH;
    return result;
  }
  
  //////////////////
  // Helping methods
  
  String getDecimalSeparator () {
    Locale locale;
    if( SessionLocale.isSet() ) {
      locale = SessionLocale.get();
    } else {
      locale = Locale.getDefault();
    }
    NumberFormat numberFormat = NumberFormat.getCurrencyInstance( locale );
    DecimalFormat format = ( DecimalFormat )numberFormat;
    DecimalFormatSymbols formatSymbols = format.getDecimalFormatSymbols();
    char decimalSeparator = formatSymbols.getDecimalSeparator();
    return String.valueOf( decimalSeparator );
  }

  private static int checkStyle( final int style ) {
    /*
    * Even though it is legal to create this widget
    * with scroll bars, they serve no useful purpose
    * because they do not automatically scroll the
    * widget's client area.  The fix is to clear
    * the SWT style.
    */
    return style & ~( RWT.H_SCROLL | RWT.V_SCROLL );
  }
}
